### DISCUSSION TITLE
64 bit integers

### DISCUSSION THREAD

**Discussion by hros** [fastapi/fastapi#6142] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.get("/")
def read_root():
    return {f"{2**61}": 2**61}
```

### Description

trying to return a long integer value (e.g., 2**61)
numbers are rounded to next 1,000

the output of the code is:
```python
{
  "2305843009213693952": 2305843009213694000
}
```

### Operating System

Linux

### Operating System Details

Ubuntu 22.04 (in WSL2 / Windows-11)

### FastAPI Version

0.85.0

### Python Version

3.9.13

**Comment by acidjunk** [comment]
Seems double with #1511

**Comment by acidjunk** [comment]
There is also a limitation in JSON itself; more info here ->
https://github.com/tiangolo/fastapi/issues/3028

**Comment by hros** [comment]
Both of the issues mentined are referring to big-int values
I was sending a long (using less than 64 bits)
Afaik openapi/swagger supports long (64 bit) types

**Comment by JarroVGIT** [comment]
How are you calling your endpoint? Because I do think you are running into the issues mentioned by @acidjunk . 

Point in case, when I run your code and call the endpoint with curl, the result is this:

```zsh
jarro@MBP-van-Jarro fastapi-github-issues % curl localhost:8000
{"2305843009213693952":2305843009213693952}%  
```

However, when parsed in a browser (which has JSON extension turned on for prettifying json):

<img width="801" alt="image" src="https://user-images.githubusercontent.com/13659033/194402781-eac4165e-00fe-466f-8eb8-23bf16907d04.png">

**Comment by hros** [comment]
thanks @JarroVGIT , you are right
So, is it only a browser issue? or a JSON issue?

A couple of other questions:
1. If I understand correctly, FastAPI supports both the JSON standard and the OpenAPI standard. When is either standard used?
2. After resolving the 64-bit issue, is it possible to send/receive 128-bit values using FastAPI, or only by conversion (either serializing the data, or stringifying it)

**Comment by JarroVGIT** [comment, marked as answer]
> If I understand correctly, FastAPI supports both the JSON standard and the OpenAPI standard. When is either standard used?

These are two wildly different standards that have completely different goals. JSON is just... that, JSON. FastAPI imports the `json` package and uses that, in a lot of places where it handles JSON. The OpenAPI standard is used when generating the `openapi.json` file. 

Typically, you only have to concern yourself with JSON when developing your API and with OpenAPI when you are dealing with the generated OpenAPI schema (`openapi.json`). 

The problem you ran into, is neither a limitation of JSON nor OpenAPI, but of Javascript not being able to parse such large `int`s correctly (or to be more precise: with such precision). The screenshot I showed was where I used a JSON prettifier in the browser, which uses Javascript.

> After resolving the 64-bit issue, is it possible to send/receive 128-bit values using FastAPI, or only by conversion (either serializing the data, or stringifying it)

Yes you can; using this:
```python
@app.get("/")
def read_root():
    return {f"{2**127}": 2**127}
```
Which results in (curl):
```zsh
jarro@MBP-van-Jarro fastapi-github-issues % curl localhost:8000
{"170141183460469231731687303715884105728":170141183460469231731687303715884105728}%   
```
but in a browser (with a Javascript JSON parser active):
<img width="598" alt="image" src="https://user-images.githubusercontent.com/13659033/194475329-25541f74-9d4a-44cd-bcc0-2d42218db6fc.png">

**Comment by hros** [comment]
That's fantastic. thanks @JarroVGIT 
Naturally, I want to read the data using api, not curl
I'll try and parse 128 types with [cppRestSDK](https://github.com/Microsoft/cpprestsdk) (do you by any chance happen to know how to do that?)

**Comment by JarroVGIT** [comment]
Curl is just a tool that you can use to parse the raw output of your API. Your issue is **not** a FastAPI issue but a Javascript issue (because you probably were using the browser to go to your endpoint). So, if your client can handle large int's, then you should be fine. I have no experience with ccpRestSDK, all I know is that FastAPI will output it correctly :)

**Comment by tiangolo** [comment]
Thanks for the help here @JarroVGIT and @acidjunk  ! 👏  🙇 

Thanks for reporting back and closing the issue @hros  👍
